libraries/Microsoft.Bot.Builder.AI.LuisV3/LuisApplication.cs

```diff
@@ -0,0 +1,105 @@
+        public LuisApplication(string applicationId, string endpointKey, string endpoint)
+            : this((applicationId, endpointKey, endpoint))
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LuisApplication"/> class.
+        /// </summary>
+        /// <param name="applicationEndpoint">LUIS application endpoint.</param>
+        public LuisApplication(string applicationEndpoint)
+            : this(Parse(applicationEndpoint))
+        {
+        }
+
+        private LuisApplication(ValueTuple<string, string, string> props)
+        {
+            var (applicationId, endpointKey, endpoint) = props;
+
+            if (!Guid.TryParse(applicationId, out var _))
+            {
+                throw new ArgumentException($"\"{applicationId}\" is not a valid LUIS application id.");
+            }
+
+            if (!Guid.TryParse(endpointKey, out var subscriptionGuid))
+            {
+                throw new ArgumentException($"\"{subscriptionGuid}\" is not a valid LUIS subscription key.");
+            }
+
+            if (string.IsNullOrWhiteSpace(endpoint))
+            {
+                endpoint = "https://westus.api.cognitive.microsoft.com";
+            }
+
+            if (!Uri.IsWellFormedUriString(endpoint, UriKind.Absolute))
+            {
+                throw new ArgumentException($"\"{endpoint}\" is not a valid LUIS endpoint.");
+            }
+
+            ApplicationId = applicationId;
+            EndpointKey = endpointKey;
+            Endpoint = endpoint;
+        }
+
+        /// <summary>
+        /// Gets or sets lUIS application ID.
+        /// </summary>
+        /// <value>
+        /// LUIS application ID.
+        /// </value>
+        public string ApplicationId { get; set; }
+
+        /// <summary>
+        /// Gets or sets lUIS subscription or endpoint key.
+        /// </summary>
+        /// <value>
+        /// LUIS subscription or endpoint key.
+        /// </value>
+        public string EndpointKey { get; set; }
+
+        /// <summary>
+        /// Gets or sets lUIS endpoint like https://westus.api.cognitive.microsoft.com.
+        /// </summary>
+        /// <value>
+        /// LUIS endpoint where application is hosted.
+        /// </value>
+        public string Endpoint { get; set; }
+
+        private static (string applicationId, string endpointKey, string endpoint) Parse(string applicationEndpoint)
+        {
+            if (!Uri.TryCreate(applicationEndpoint, UriKind.Absolute, out var uri))
+            {
+                throw new ArgumentException(nameof(applicationEndpoint));
+            }
+
+            var applicationId = uri.Segments.Last();
+            var endpointKey = HttpUtility.ParseQueryString(uri.Query).Get("subscription-key");
+            var endpoint = uri.GetLeftPart(UriPartial.Authority);
+            return (applicationId, endpointKey, endpoint);
```

Comment:
Shouldn't you add a secondary Id here? The slot name in case of using slots, and the version Id in case of hitting a version.

Reply:
Those are prediction options, you can change them per query.  These are the stable part of an application.  This is also compatible with the V2 API.

Reply:
In my opinion, no one would alternate between versions and slots in different API calls. They could have totally different schemas, but ok.